### PR TITLE
Remove circular references when stringifying an object

### PR DESCRIPTION
**Pull request by hbcarlos**
Remove circular references when stringifying an object because otherwise, the `JSON.stringify` throws an error.

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks!

This seems to be working fine when testing on Binder:

image


Using a snippet like in: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#issue_with_json.stringify_when_serializing_circular_references

File: src/index.ts [hunk 1]
```diff
@@ -169,13 +169,31 @@ const extension: JupyterFrontEndPlugin<void> = {
     const _trace = console.trace;
     const _table = console.table;
 
+    // https://stackoverflow.com/a/11616993
```
Comment (jtpio) [inline src/index.ts:172]: Wondering whether we want to take this approach though.

As mentioned  in the SA answer:

> The replacer in this example is not 100% correct (depending on your definition of "duplicate"). In the following case, a value is discarded:

This indeed discards some of the elements:

image

Comment (hbcarlos) [reply src/index.ts:172]: Added a warning message informing the user about the possibility of missing attributes

**Comment by hbcarlos** [comment]
Binder link working https://mybinder.org/v2/gh/QuantStack/jupyterlab-js-logs/circular-references

**Comment by jtpio** [comment]
Nice!

File: src/index.ts [hunk 2]
```diff
@@ -291,10 +395,23 @@ const extension: JupyterFrontEndPlugin<void> = {
...
+          try {
+            const msg =
+              'This error contains a object with a circular reference. During the process of removing the reference we could have removed duplicated attributes.\n';
+            const obj = JSON.stringify(arg, refReplacer);
+            cache = [];
+            console.error(msg, obj);
+            data += obj;
+          } catch (e) {
+            data += ' ';
+          }
```
Comment (jtpio) [inline src/index.ts:413]: Maybe we could do a quick refactor if this piece of code is the same in all cases?

Comment (hbcarlos) [reply src/index.ts:413]: ooh, thanks. I forgot about that.
I'm on it.

File: src/index.ts [hunk 3]
```diff
@@ -169,138 +169,111 @@ const extension: JupyterFrontEndPlugin<void> = {
...
-            : arg) + ' ';
+        try {
+          data +=
+            (typeof arg === 'object' && arg !== null
+              ? JSON.stringify(arg)
+              : arg) + ' ';
+        } catch (e) {
+          try {
+            const msg =
+              'This error contains a object with a circular reference. During the process of removing the reference we could have removed duplicated attributes.\n';
```
Comment (jtpio) [inline src/index.ts:196]:
```suggestion
              'This error contains a object with a circular reference. Duplicated attributes might have been dropped during the process of removing the reference.\n';
```

**Review by jtpio — APPROVED** [review]
Thanks!

**Comment by jtpio** [comment]
Let's wait for CI and a quick check on Binder and this should be good to go.

**Comment by jtpio** [comment]
Binder seems to be having issues right now.

Just pulled locally and it looks good :+1: 

image

**Pull request merged**